test/e2e/e2e_test.go

```diff
@@ -497,6 +503,90 @@ var _ = ginkgo.Describe("JobSet", func() {
 			})
 		})
 	})
+	ginkgo.It("should ensure the metrics endpoint is serving metrics", func() {
+		serviceAccountName := "jobset-controller-manager"
+		metricsServiceName := "jobset-controller-manager-metrics-service"
+		namespace := "jobset-system"
+		var controllerPodName string
+
+		ginkgo.By("fetching the controller pod name")
+		cmd := exec.Command("kubectl", "get",
+			"pods", "-l", "control-plane=controller-manager",
+			"-n", namespace,
+			"-o", "go-template={{ range .items }}"+
+				"{{ if not .metadata.deletionTimestamp }}"+
+				"{{ .metadata.name }}"+
+				"{{ \"\\n\" }}{{ end }}{{ end }}",
+		)
+
+		podOutput, err := testutils.Run(cmd)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred(), "Failed to retrieve controller-manager pod information")
+		podNames := testutils.GetNonEmptyLines(podOutput)
+		controllerPodName = podNames[0]
+
+		ginkgo.By("validating that the metrics service is available")
+		cmd = exec.Command("kubectl", "get", "service", metricsServiceName, "-n", namespace)
+		_, err = testutils.Run(cmd)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred(), "Metrics service should exist")
+
+		ginkgo.By("getting the service account token")
+		token, err := serviceAccountToken(serviceAccountName, namespace)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		gomega.Expect(token).NotTo(gomega.BeEmpty())
+
+		ginkgo.By("waiting for the metrics endpoint to be ready")
+		verifyMetricsEndpointReady := func(g gomega.Gomega) {
+			cmd := exec.Command("kubectl", "get", "endpoints", metricsServiceName, "-n", namespace)
+			g.Expect(err).NotTo(gomega.HaveOccurred())
+			output, err := testutils.Run(cmd)
+			g.Expect(err).Should(gomega.BeNil())
+			g.Expect(output).To(gomega.ContainSubstring("8443"), "Metrics endpoint is not ready")
+		}
+		gomega.Eventually(verifyMetricsEndpointReady).Should(gomega.Succeed())
+
+		ginkgo.By("verifying that the controller manager is serving the metrics server")
+		verifyMetricsServerStarted := func(g gomega.Gomega) {
+			cmd := exec.Command("kubectl", "logs", controllerPodName, "-n", namespace)
+			output, err := testutils.Run(cmd)
+			g.Expect(err).NotTo(gomega.HaveOccurred())
+			g.Expect(output).To(gomega.ContainSubstring("controller-runtime.metrics\tServing metrics server"),
+				"Metrics server not yet started")
+		}
+		gomega.Eventually(verifyMetricsServerStarted).Should(gomega.Succeed())
+
+		ginkgo.By("creating the curl-metrics job to access the metrics endpoint")
+		cmd = exec.Command("kubectl", "create", "job", "curl-metrics",
+			"--namespace", namespace,
+			"--image=quay.io/openshifttest/hello-openshift:1.2.0",
```

Comment:
In upstream, i think we cannot use private image, maybe use like lws test: https://github.com/kubernetes-sigs/lws/blob/main/test/e2e/e2e_test.go#L371C14-L371C23